File: exercises/stack_cpp/stack.cpp
#include <iostream>
#include <vector>
#include <string>

class Stack
{
private:
    std::vector<std::string> stack;
public:
    Stack(/* args */);
    ~Stack();

    void push(std::string);
    std::string pop();
    bool is_empty();
    std::string peek();
    int size();
};

Stack::Stack(/* args */)
{
}

Stack::~Stack()
{
}

void Stack::push(std::string x)
{
    stack.push_back(x);
}

std::string Stack::pop()
{
    std::string tmp = stack.back();
    stack.pop_back();
    return tmp;
}

bool Stack::is_empty()
{
    return stack.size() == 0;
}

std::string Stack::peek()
{
    return stack.back();
}

int Stack::size()
{
    return stack.size();
}

int main()
{
    Stack myStack;
    std::cout << std::boolalpha;
    std::cout << "Empty? " << myStack.is_empty() << "\n";
    std::cout << "Pushing \"3\"\n";
    myStack.push("3");
    std::cout << "Pushing \"2\"\n";
    myStack.push("2");
    std::cout << "Pushing \"1\"\n";
    myStack.push("1");
    std::cout << "Peeking: " << myStack.peek() << "\n";
    std::cout << "Size? " << myStack.size() << "\n";
    std::cout << "Popping: " << myStack.pop() << "\n";
    std::cout << "Popping: " << myStack.pop() << "\n";
    std::cout << "Empty? " << myStack.is_empty() << "\n";
    std::cout << "Size? " << myStack.size() << "\n";
    std::cout << "Popping: " << myStack.pop() << "\n";
    std::cout << "Empty? " << myStack.is_empty() << "\n\n";
    return 0;
}

File: exercises/stack_py/stack.py
class Stack:
    def __init__(self):
        self.stack = []
        self._size = 0
    
    def push(self, x):
        self.stack.insert(0, x)
        self._size += 1
    
    def pop(self):
        self._size -= 1
        return self.stack.pop(0)
    
    def is_empty(self):
        return self._size == 0
    
    def peek(self):
        if self._size == 0:
            return ""
        else:
            return self.stack[0]
    
    def size(self):
        return self._size


if __name__ == '__main__':
    myStack = Stack()
    print(f"Empty? {myStack.is_empty()}")
    print("Pushing 3")
    myStack.push(3)
    print("Pushing 2")
    myStack.push(2)
    print("Pushing 1")
    myStack.push(1)
    print(f"Peeking: {myStack.peek()}")
    print(f"Size? {myStack.size()}")
    print(f"Popping: {myStack.pop()}")
    print(f"Popping: {myStack.pop()}")
    print(f"Empty? {myStack.is_empty()}")
    print(f"Size? {myStack.size()}")
    print(f"Popping: {myStack.pop()}")
    print(f"Empty? {myStack.is_empty()}")